File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/team/supervisor_test.go
package team

import (
	"strings"
	"testing"
)

func TestParsesDelegationDecision(t *testing.T) {
	decision, err := ParseDecision(`{"next":"analista","brief":"diagnostique"}`)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if decision.Done {
		t.Fatal("expected not done")
	}
	if decision.Next != RoleAnalista {
		t.Fatalf("expected analista, got %s", decision.Next)
	}
	if decision.Brief != "diagnostique" {
		t.Fatalf("unexpected brief: %s", decision.Brief)
	}
}

func TestParsesDoneDecision(t *testing.T) {
	decision, err := ParseDecision(`{"next":"done","brief":"resumo final"}`)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !decision.Done {
		t.Fatal("expected done")
	}
	if decision.Brief != "resumo final" {
		t.Fatalf("unexpected brief: %s", decision.Brief)
	}
}

func TestMalformedJSONErrors(t *testing.T) {
	_, err := ParseDecision("nao e json")
	if err == nil || !strings.Contains(err.Error(), "decisao invalida do supervisor") {
		t.Fatalf("expected 'decisao invalida do supervisor' error, got %v", err)
	}
}

func TestUnknownRoleErrors(t *testing.T) {
	_, err := ParseDecision(`{"next":"estagiario","brief":"x"}`)
	if err == nil || !strings.Contains(err.Error(), "papel desconhecido") {
		t.Fatalf("expected 'papel desconhecido' error, got %v", err)
	}
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/test/java/com/opspilot/SupervisorTest.java
package com.opspilot;

import com.opspilot.team.SupervisorDecision;
import com.opspilot.team.TeamRole;
import org.junit.jupiter.api.Test;

import static com.opspilot.team.Supervisor.parseDecision;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class SupervisorTest {

    @Test
    void parsesDelegationDecision() {
        SupervisorDecision decision = parseDecision("{\"next\":\"analista\",\"brief\":\"diagnostique\"}");

        assertThat(decision.isDone()).isFalse();
        assertThat(decision.next()).isEqualTo(TeamRole.ANALISTA);
        assertThat(decision.brief()).isEqualTo("diagnostique");
    }

    @Test
    void parsesDoneDecision() {
        SupervisorDecision decision = parseDecision("{\"next\":\"done\",\"brief\":\"resumo final\"}");

        assertThat(decision.isDone()).isTrue();
        assertThat(decision.brief()).isEqualTo("resumo final");
    }

    @Test
    void malformedJsonThrows() {
        assertThatThrownBy(() -> parseDecision("nao e json"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("decisao invalida do supervisor");
    }

    @Test
    void unknownRoleThrows() {
        assertThatThrownBy(() -> parseDecision("{\"next\":\"estagiario\",\"brief\":\"x\"}"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("papel desconhecido");
    }
}
